File: Python/144.Binary Tree Preorder Traversal.py
#先序遍历一个二叉树
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        self.result = []
        self.preorder(root)
        return self.result

    def preorder(self,p):
        if p!=None:
            self.result.append(p.val)
            self.preorder(p.left) if p.left!=None else None
            self.preorder(p.right) if p.right!=None else None


t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t1.left = t2
t1.right = t3
t3.left = t4
t4.right = t5
t2.right = t6

s = Solution()
print(s.preorderTraversal(t1))

File: C++/Stack/144.Binary Tree Preorder Traversal.cpp
/*
前序遍历二叉树。
*/
/*
思路：使用栈，在压栈前将当前结点遍历。
*/
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> preorder;
        stack<TreeNode*> mystack;
        while(root || !mystack.empty())
        {
        	while(root)
        	{
        		preorder.push_back(root->val);
        		mystack.push(root);
        		root = root->left;
        	}
        	if(!mystack.empty())
        	{
        		root = mystack.top()->right;
        		mystack.pop();
        	}
        }
        return preorder;
    }
};
int main()
{
	Solution s;
	TreeNode root(1),t1(2),t2(3);
	root.right = &t1;
	t1.left = &t2;
	vector<int> out = s.preorderTraversal(&root);
	for(int n:out)
		cout<<n<<" ";
	cout<<endl;
}

